kiva/blend2d.py

```diff
@@ -0,0 +1,780 @@
+            spread_method, blend2d.ExtendMode.PAD
+        )
+        for stop in stops:
+            gradient.add_stop(stop[0], stop[1:])
+        self.gc.set_fill_style(gradient)
+
+    def radial_gradient(self, cx, cy, r, fx, fy, stops, spread_method,
+                        units="userSpaceOnUse"):
+        """ Sets a radial gradient as the current brush.
+        """
+        gradient = blend2d.RadialGradient(cx, cy, fx, fy, r)
+        gradient.extend_mode = gradient_spread_modes.get(
+            spread_method, blend2d.ExtendMode.PAD
+        )
+        for stop in stops:
+            gradient.add_stop(stop[0], stop[1:])
+        self.gc.set_fill_style(gradient)
+
+    # ----------------------------------------------------------------
+    # Drawing Images
+    # ----------------------------------------------------------------
+
+    def draw_image(self, img, rect=None):
+        """
+        img is either a N*M*3 or N*M*4 numpy array, or a PIL Image
+
+        rect - a tuple (x, y, w, h)
+        """
+        from PIL import Image
+
+        def normalize_image(img):
+            if not img.mode.startswith("RGB"):
+                img = img.convert("RGB")
+            return img
+
+        if isinstance(img, np.ndarray):
+            # Numeric array
+            img = Image.fromarray(img)
+            img = normalize_image(img)
+            img_array = np.array(img)
+        elif isinstance(img, Image.Image):
+            img = normalize_image(img)
+            img_array = np.array(img)
+        elif isinstance(img, GraphicsContext):
+            img_array = img._buffer
+        elif hasattr(img, "bmp_array"):
+            # An offscreen kiva.agg context
+            # XXX: Use a copy to kill the read-only flag which plays havoc
+            # with the Cython memoryviews used by blend2d
+            img = Image.fromarray(img.bmp_array)
+            img = normalize_image(img)
+            img_array = np.array(img)
+        else:
+            msg = "Cannot render image of type '{}' into blend2d context."
+            warnings.warn(msg.format(type(img)))
+            return
+
+        # XXX: Upside down!
+        dst_rect = blend2d.Rect(*rect)
+        w, h = img.width, img.height
+        image = blend2d.Image(img_array)
+        rect = blend2d.Rect(0, 0, w, h)
+        self.gc.blit_scaled_image(dst_rect, image, rect)
+
+    # ----------------------------------------------------------------
+    # Drawing Text
+    # ----------------------------------------------------------------
+
+    def select_font(self, face_name, size=12, style="regular", encoding=None):
+        """ Set the font for the current graphics context.
+        """
+        self.set_font(Font(face_name, size=size, style=style))
+
+    def set_font(self, font):
+        """ Set the font for the current graphics context.
+        """
+        spec = font.findfont()
+        self._kiva_font = font
+        self.font = blend2d.Font(spec.filename, font.size)
```

Comment:
dunno if `blend2d.Font` supports `family` which is being introduced in a separate PR.

Reply:
It's OK, it's not relevant here. I think `family` is only useful for APIs which handle the font lookup for us.